### PR TITLE
[master] Fix Python3.13 compatibility regarding urllib.parse module

### LINKED ISSUE #66898: [TECH DEBT] Salt fails to load top.sls on Python3.13rc2
### Description of the tech debt to be addressed, include links and screenshots

A recent change in Python 3.13 causes salt to fails on loading top.sls. With debugging enabled, I see now:
```
[DEBUG   ] Could not find file 'salt://.sls' in saltenv 'base'
[DEBUG   ] No contents loaded for saltenv 'base'
```

This seems to be directly caused by the change in Python: https://github.com/python/cpython/issues/85110

Old behavior:
```
>>> urllib.parse.urlunparse(("file", "", "top.sls", "", "", ""))
'file:///top.sls'
```
New behavior:
```
>>> urllib.parse.urlunparse(("file", "", "top.sls", "", "", ""))
'file:top.sls'
```

Technically, I think the latter might be more correct. But it still breaks salt's expectations. Specifically here: https://github.com/saltstack/salt/blob/246d0664577ef72da8bd1f0c4dff0d18b4428b23/salt/utils/url.py#L49-L50

With `file:top.sls`, looking for length of `file:///` is wrong now.

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)

```
Salt Version:
          Salt: 3006.9
 
Python Version:
        Python: 3.13.0rc2 (main, Sep  7 2024, 00:00:00) [GCC 14.2.1 20240801 (Red Hat 14.2.1-1)]
 
Dependency Versions:
          cffi: 1.17.0
      cherrypy: Not Installed
  cryptography: Not Installed
      dateutil: 2.8.2
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.4
       libgit2: Not Installed
  looseversion: 1.3.0
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.6
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 24.1
     pycparser: 2.20
      pycrypto: Not Installed
  pycryptodome: 3.20.0
        pygit2: Not Installed
  python-gnupg: Not Installed
        PyYAML: 6.0.1
         PyZMQ: 25.1.1
        relenv: Not Installed
         smmap: Not Installed
       timelib: Not Installed
       Tornado: Not Installed
           ZMQ: 4.3.5
 
System Versions:
          dist: qubes 4.3 R4.3
        locale: utf-8
       machine: x86_64
       release: 6.6.48-1.qubes.fc41.x86_64
        system: Linux
       version: Qubes OS 4.3 R4.3
```

The same issue applies to Fedora 41 (in beta now). And while I tested using Salt 3006.9, the relevant code is the same in master.

### PR DESCRIPTION
**Pull request by marmarek**
### What does this PR do?

Python 3.13 fixed handling relative paths in `urllib.parse` module. Specifically, relative file URL is now constructed as `file:path` instead of converting it to absolute `file:///path`. This breaks `salt.utils.url.create` which expects `file:///` specifically. The mismatch results in for example changing `salt://top.sls` into `salt://.sls` and thus not finding the top file.

Fix this by handling both prefixes.

Relevant python change: https://github.com/python/cpython/issues/85110

### What issues does this PR fix or reference?
Fixes: #66898

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated (current tests in tests/unit/utils/test_url.py are enough when will start running on Python 3.13)

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices, including the
[PR Guidelines](https://docs.saltproject.io/en/master/topics/development/pull_requests.html).

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Review by twangboy — CHANGES_REQUESTED** [review]
Could you please write a test for this

**Review by dmurphy18 — CHANGES_REQUESTED** [review]
Needs tests written in pytest

**Comment by marmarek** [comment]
> Could you please write a test for this

The existing tests/unit/utils/test_url.py already covers this. And indeed, without my change, it fails already on Python 3.13rc2 (Fedora 41 beta) or Python 3.12.6 (Fedora 40). The change that broke did got backported to Python 3.12.6 (released about 2 weeks ago). I see you have Fedora 40 in CI already, but it seems to run use onedir (and Python 3.10.x inside) instead of Python from Fedora. Is there any CI job that tests on Python 3.12.x (or 3.13)?

**Comment by twangboy** [comment]
We currently only support up to Python 3.10. There are dependency issues we need to get figured out across all the Operating systems we support before we can officially support anything newer.

**Comment by marmarek** [comment]
In absence of newer Python to actually test the behavior there, do you want me to mimic its new behavior by patching `urllib.parse.urlunparse` then (I find it weird, but if you insist I can add it)? The code is already tested that it still works on Python 3.10. And once you will be adding support for Python 3.12, you will have one thing less to fix :)

BTW, does the CI report somewhere test coverage? From what I can see, coverage-related jobs got skipped...

**Review by slimlv — CHANGES_REQUESTED** [review]
how many "If/return"s  you going to insert ?
note that changes in python code was introduced in minor release 3.12.6

try to make smart bulletproof improvements , for example
```python
url = salt.utils.data.decode(urlunparse(("file", "", "/" + path, "", query, "")))

**Comment by twangboy** [comment]
I think, just a unit test to make sure that if the URL starts with `file:///` it gets converted to 'salt://`

File: salt/utils/url.py [hunk 1]
```diff
@@ -47,7 +47,10 @@ def create(path, saltenv=None):
 
     query = f"saltenv={saltenv}" if saltenv else ""
     url = salt.utils.data.decode(urlunparse(("file", "", path, "", query, "")))
-    return "salt://{}".format(url[len("file:///") :])
+    # urlunparse changed behavior in Python 3.13
+    if url.startswith("file:///"):
+        return "salt://{}".format(url[len("file:///") :])
+    return "salt://{}".format(url[len("file:") :])
```
Comment (bdrx312) [inline salt/utils/url.py:53]: This create function should not be doing `path = path.replace("\\", /)` and probably not 
```python
if salt.utils.platform.is_windows():
path = salt.utils.path.sanitize_win_path(path)
```
Instead it should use `urllib.request.pathname2url` or some other python builtin functionality.

Especially since the `sanitize_win_path` function is clearly miswritten:

```python
def sanitize_win_path(winpath):
    """
    Remove illegal path characters for windows
    """
    intab = "<>:|?*"
    if isinstance(winpath, str):
        winpath = winpath.translate({ord(c): "_" for c in intab})
    elif isinstance(winpath, str):
```
Notice the duplicate `isinstance(winpath, str)` in the elif that will never occur.

I would prefer instead of assuming that the string starts with `file:` or `file:///` we do a regex search replace like `re.sub(r"^file:(?:///)?", "", url)` Also use `f` string instead of `format`.

```python
return f'salt://{re.sub(r"^file:(?:///)?", "", url)}'
```

```

Comment (slimlv) [reply salt/utils/url.py:53]: `return f'salt://{re.sub(r"^file:(?:/+)?", "", url)}'`

Comment (bdrx312) [reply salt/utils/url.py:53]: > ```python
> return f'salt://{re.sub(r"^file:(?:/+)?", "", url)}'
> ```

If you are going to do zero or more you don't `+` and `?` you just use `*`

```python
return f'salt://{re.sub(r"^file:(?:/*)", "", url)}'
```

Comment (bdrx312) [reply salt/utils/url.py:53]: and if you want to do 0 to 3 `/` then it would be 
```python
return f'salt://{re.sub(r"^file:/{0,3}", "", url)}'
```

or 

```python
return f'salt://{re.sub(r"^file:/{,3}", "", url)}'
```
but I think the explicit `0` is more readable.

**Comment by marmarek** [comment]
> I think, just a unit test to make sure that if the URL starts with `file:///` it gets converted to 'salt://`

Most tests in tests/unit/utils/test_url.py do that already, for example `test_create_url`. Those tests do fail when running on Python 3.13 (or Python 3.12.6), so this regression is covered already. Your CI will notice it as soon as you start testing on such Python version.
And BTW, the `file://` prefix is added only in `salt.utils.url.create()`, just to be replaced with `salt://` - it doesn't exist in the function call argument.

File: salt/utils/url.py [hunk 2]
```diff
@@ -47,7 +47,7 @@ def create(path, saltenv=None):
 
     query = f"saltenv={saltenv}" if saltenv else ""
     url = salt.utils.data.decode(urlunparse(("file", "", path, "", query, "")))
-    return "salt://{}".format(url[len("file:///") :])
+    return f'salt://{re.sub(r"file:/*", "", url)}'
```
Comment (bdrx312) [inline salt/utils/url.py:50]: The regex needs a begin anchor `^` so it doesn't replace `file:/` somewhere in the middle of the url: `f'salt://{re.sub(r"^file:/*", "", url)}'

**Review by bdrx312 — APPROVED** [review]

File: salt/utils/url.py [hunk 3]
```diff
@@ -47,7 +47,7 @@ def create(path, saltenv=None):
 
     query = f"saltenv={saltenv}" if saltenv else ""
     url = salt.utils.data.decode(urlunparse(("file", "", path, "", query, "")))
```
Comment (yangskyboxlabs) [inline salt/utils/url.py:49]: I recommend just not using schema when calling `unparse`/`unsplit`:

```suggestion
   return f"salt://{salt.utils.data.decode(urlunsplit(("", "", path, query, "")))}"
```

Comment (bdrx312) [reply salt/utils/url.py:49]: I just read through the source code of `urlunparse`/`urlunsplit` and it looks like there is not currently any special handling for the file scheme so I agree that this is probably a better solution.

**Comment by jfindlay** [comment]
In my opinion, having written both the original `salt.utils.url` and its accompanying tests, I agree the existing test cases already cover the code well enough.  I don't think it is necessary to cover the case that upstream will again change the behavior of code that hasn't been modified in ~30 years again.

It is a subtle breaking change for a use case they very likely didn't expect.  Regexing `file:///` for `salt://` is a yolo move, whereas subclassing the urlparse code to add a `salt` proto would have been more proper.  (Maybe the upstream code doesn't care because you already can?)  If anything a rare bug in old code is an indictment of the poor usecase.  Not that `urlparse.urlunparse`'s api is user friendly (pass the url components in a fixed length anonymous list; give me a break).  There are still many occurrences in the codebase that handcraft `salt://` URLs anyway.

Ironically, many years ago when I worked on this, Tom said "I wish we had gone with `salt:///`.

File: salt/utils/url.py (hunk 3 above)
Comment (marmarek) [reply salt/utils/url.py]: Ok, lets see how this works. Tests on Python 3.12.6 are green with this version, I'll let CI check older ones.

Comment (bdrx312) [reply salt/utils/url.py]: Using urlunparse is okay, but I think using `urlunsplit` as @yangskyboxlabs is more ideal since it passes one less unused argument.

**Review by bdrx312 — APPROVED** [review]

**Comment by dead10ck** [comment]
I'd like to remind the maintainers that this bug completely and utterly breaks salt on many systems, especially because this "fix" has been backported to Python 3.12.

I understand that 3.10 is currently the officially supported version, but the reality is that a lot of folks are not, and never will, use the onedir package, so this is high impact regardless of official compatibility guarantees.

I understand the desire for tests with changes in the general case, but I think folks have to keep in mind that that should be a rule of thumb, not a hard requirement in every case no matter what the cost. Sometimes fixing high impact bugs, level of effort to benefit ratio, and potential risk for introducing new regressions in more strongly supported environments should weigh into the decision to block merging unless there are tests. It's been established that there is significant preexisting coverage on this code path, so blocking on adding more tests is not doing anyone any good.

**Review by twangboy — APPROVED** [review]

**Review by bdrx312 — APPROVED** [review]

**Comment by hartwork** [comment]
FYI Fedora has integrated this patch into their packaging by now at https://src.fedoraproject.org/rpms/salt/blob/main/f/urllib.patch :clap:

**Review by twangboy — CHANGES_REQUESTED** [review]
Is there a way to write a test for this?

**Comment by marmarek** [comment]
> Is there a way to write a test for this?

See https://github.com/saltstack/salt/pull/66899#issuecomment-2359371795
tl;dr: existing tests cover this extensively already. But CI doesn't run Python 3.13 tests (yet?).

**Comment by twangboy** [comment]
Ah, true

**Review by twangboy — APPROVED** [review]

**Comment by printerderby** [comment]
I'd love to see this fix in 3007, as `salt-ssh` minions use the system's version of python. #68080
Every newer distribution is causing troubles with `file.managed` and `salt://`-URLs. E.g.: Debian Trixie, Alpine Linux, Arch Linux, Ubuntu 24.04, etc.

**Comment by twangboy** [comment]
If this is a bug on an older branch, then the PR should be created against the oldest supported branch where the bug exists.

**Comment by pjcreath** [comment]
This affects 3006.14 as well as Python 3.12.6 and later.

File: salt/utils/url.py [hunk 4]
```diff
@@ -46,8 +46,7 @@ def create(path, saltenv=None):
     path = salt.utils.data.decode(path)
 
     query = f"saltenv={saltenv}" if saltenv else ""
-    url = salt.utils.data.decode(urlunparse(("file", "", path, "", query, "")))
-    return "salt://{}".format(url[len("file:///") :])
+    return f'salt://{salt.utils.data.decode(urlunsplit(("", "", path, query, "")))}'
```
Comment (bdrx312) [inline salt/utils/url.py:49]: Apparently there are use cases where `file:///` gets passed in as the prefix on `path` and the old logic would strip it off. It seems incorrect that places call this method with `file://` still on the path, but to avoid breaking those cases this method should handle removing it like it previously did. We should add that use case to the test procedures.

Comment (twangboy) [reply salt/utils/url.py:49]: Yes, we need to handle `file:///` paths

**Comment by dwoz** [comment]
@bdrx312 @twangboy 

Is everything ready here?

@pjcreath 

Assuming this get's merged here. If you would like to back port this to 3007.x or even 3006.x we'll merge it.

**Comment by bdrx312** [comment]
> @bdrx312 @twangboy
> 
> Is everything ready here?

I think this https://github.com/saltstack/salt/pull/66899#discussion_r2323770702 still needs to be handled and is not. I vaguely remember reading in another issue or PR that someone tried this solution and there was a use case that broke because it didn't handle the file:/// paths with the same return value. It was so long ago that I posted I don't remember the exact details at the moment.

File: salt/utils/url.py (hunk 4 above)
Comment (l33) [reply salt/utils/url.py:49]: > Apparently there are use cases where `file:///` gets passed in as the prefix on `path` and the old logic would strip it off.

How so? The old code only strips off a fixed number of characters and urlunparse still adds the prefix even if it is already part of the path:

```
>>> urlunparse(("file", "", "file:///path", "", "query", ""))
'file:///file:///path?query'
>>> urlunparse(("file", "", "file:///path", "", "query", ""))[len("file:///") :]
'file:///path?query'
```

Comment (bdrx312) [reply salt/utils/url.py:49]: @l33 I don't know why I didn't link the comment in the other issue when I read it, but I should have. I don't remember now, and it appears I was wrong about the cause according to your investigation.

Comment (bdrx312) [reply salt/utils/url.py:49]: @dwoz It appears this is not an issue, so this should be ready. Clicking resolve on this comment fails.

Comment (lkubb) [reply salt/utils/url.py:49]: @bdrx312 I think you're referring to this comment: https://github.com/saltstack/salt/issues/68267#issuecomment-3192118739

Indeed, the old code would add and remove a `file:///` prefix, so this should not be an issue.

**Comment by twangboy** [comment]
The has been resolved and can be merged.
